rust/protocol/src/protocol.rs

```diff
@@ -403,29 +413,63 @@ impl SenderKeyMessage {
             ciphertext: Some(ciphertext.to_vec()),
         };
         let proto_message_len = proto_message.encoded_len();
-        let mut serialized = vec![0u8; 1 + proto_message_len + Self::SIGNATURE_LEN];
+        let mut serialized = vec![0u8; 1 + proto_message_len + SIGNATURE_LENGTH];
         serialized[0] = ((message_version & 0xF) << 4) | SENDERKEY_MESSAGE_CURRENT_VERSION;
         proto_message.encode(&mut &mut serialized[1..1 + proto_message_len])?;
         let signature =
             signature_key.calculate_signature(&serialized[..1 + proto_message_len], csprng)?;
         serialized[1 + proto_message_len..].copy_from_slice(&signature[..]);
+
+        let (semi_serialized, serialized, signature) =
+            Self::split_serialized_signature(&serialized)?;
+
         Ok(Self {
             message_version: SENDERKEY_MESSAGE_CURRENT_VERSION,
             distribution_id,
             chain_id,
             iteration,
             ciphertext,
-            serialized: serialized.into_boxed_slice(),
+            semi_serialized,
+            signature,
+            serialized,
         })
     }
 
-    pub fn verify_signature(&self, signature_key: &PublicKey) -> Result<bool> {
-        let valid = signature_key.verify_signature(
-            &self.serialized[..self.serialized.len() - Self::SIGNATURE_LEN],
-            &self.serialized[self.serialized.len() - Self::SIGNATURE_LEN..],
-        )?;
+    /// Return a tuple containing:
+    /// 1. the serialized message *without* the signature at the end.
+    /// 2. the complete serialized message *with* signature.
+    /// 3. the signature as a fixed-size array.
+    fn split_serialized_signature(
```

Comment:
It's bad enough that we're keeping both the serialized and unserialized forms of most of these messages; we definitely should not be copying them *again* just to have fixed-sized arrays. I think this won't matter when you put back the variable-length APIs, but in general this would be a good place for `array_ref!` or similar to reference bytes out of the buffer rather than copying it more than it already is.